pkg/operator/controller/controller.go

```diff
@@ -326,6 +313,76 @@ func (r *reconciler) ensureInternalRouterServiceForIngress(deployment *appsv1.De
 	return nil
 }
 
+// ensureRouterNamespaceAsset ensures that the router namespace exists and
+// matches the expected namespace.
+func (r *reconciler) ensureRouterNamespaceAsset() error {
+	ns, err := r.ManifestFactory.RouterNamespace()
+	if err != nil {
+		return fmt.Errorf("failed to build router namespace: %v", err)
+	}
+
+	// Need to clone namespace as `ensureRouterAsset` will create a new
+	// router asset if it doesn't exist but if the asset already exists, it
+	// will return us the modified asset (current version from api service).
+	current := ns.DeepCopy()
+	key := types.NamespacedName{Name: ns.Name}
+
+	if _, err := r.ensureRouterAsset(key, current); err != nil {
+		return err
+	}
+
+	// TODO: should we check for all label/annotation changes or just
+	//       specific ones node-selector/run-level?
+	modified := !reflect.DeepEqual(current.Labels, ns.Labels) || !reflect.DeepEqual(current.Annotations, ns.Annotations)
+	modified = modified || !reflect.DeepEqual(current.Spec, ns.Spec)
+	if modified {
+		if err := r.Client.Update(context.TODO(), ns); err != nil {
+			return err
+		}
+
+		logrus.Infof("updated router namespace %s", ns.Name)
+		return nil
+	}
+
+	return nil
+}
+
+// ensureRouterDeployment ensures that the router deployment exists and matches
+// the expected deployment.
+func (r *reconciler) ensureRouterDeployment(ci *ingressv1alpha1.ClusterIngress) (*appsv1.Deployment, error) {
+	deployment, err := r.ManifestFactory.RouterDeployment(ci)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build router deployment: %v", err)
+	}
+
+	// Need to clone the deployment as `ensureRouterAsset` will create a new
+	// router asset if it doesn't exist but if the asset already exists, it
+	// will return us the modified asset (current version from api service).
+	current := deployment.DeepCopy()
+	key := types.NamespacedName{Namespace: deployment.Namespace, Name: deployment.Name}
+
+	if _, err := r.ensureRouterAsset(key, current); err != nil {
+		return nil, err
+	}
+
+	// TODO: should we check for label/annotation changes or ignore them?
+	modified := !reflect.DeepEqual(current.Labels, deployment.Labels) || !reflect.DeepEqual(current.Annotations, deployment.Annotations)
+
+	// TODO: this will need changes when unsupported extensions/patches get
+	//       added to the router deployment.
+	modified = modified || !reflect.DeepEqual(current.Spec, deployment.Spec)
+	if modified {
+		if err := r.Client.Update(context.TODO(), deployment); err != nil {
+			return deployment, fmt.Errorf("failed to update router deployment %s, %v", key.String(), err)
```

Comment:
Returning `deployment` instead of `nil` with an error value looks odd.  The caller isn't going to use the deployment value if the error value is non-nil, so why not return `nil` for the deployment value?

Reply:
ack will do.